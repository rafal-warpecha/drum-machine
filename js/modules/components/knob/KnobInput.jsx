import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Knob.css';

export default class KnobInput extends Component {
    static propTypes = {
        onChange: PropTypes.func.isRequired,
        min: PropTypes.number.isRequired,
        max: PropTypes.number.isRequired,
        value: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]).isRequired
    };

    constructor() {
        super();

        this.onInputChange = this.onInputChange.bind(this);
        this.onInputBlur = this.onInputBlur.bind(this);
    }

    onInputChange(event) {
        const { min, max, onChange } = this.props;

        let { value } = event.target;

        if (value === '') {
            value = '';
        } else {
            value = Math.round(Math.max(Math.min(value, max), min));
        }

        onChange(value);
    }

    onInputBlur(event) {
        const { min, onChange } = this.props;

        if (event.target.value === '') {
            onChange(min);
        }
    }

    render() {
        const { min, max, value } = this.props;

        return (
            <input
                className={styles.input}
                type="number"
                value={value}
                min={min}
                max={max}
                onChange={this.onInputChange}
                onBlur={this.onInputBlur}
            />
        );
    }
}
