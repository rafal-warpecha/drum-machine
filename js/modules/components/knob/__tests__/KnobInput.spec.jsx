import React from 'react';
import { shallow } from 'enzyme';
import KnobInput from './../KnobInput';

const props = {
    onChange: () => {},
    value: 0,
    min: 0,
    max: 100
};

describe('KnobInput component', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(<KnobInput {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should set value', () => {
        const onChangeSpy = spyOn(props, 'onChange');
        const wrapper = shallow(<KnobInput {...props} />);

        wrapper.simulate('change', { target: { value: '50' } });

        expect(onChangeSpy).toHaveBeenCalledWith(50);
    });

    it('should set empty string when user delete input value', () => {
        const onChangeSpy = spyOn(props, 'onChange');
        const wrapper = shallow(<KnobInput {...props} />);

        wrapper.simulate('change', { target: { value: '' } });

        expect(onChangeSpy).toHaveBeenCalledWith('');
    });

    it('should set min value when user delete input value', () => {
        const onChangeSpy = spyOn(props, 'onChange');
        const wrapper = shallow(<KnobInput {...props} />);

        wrapper.simulate('blur', { target: { value: '' } });

        expect(onChangeSpy).toHaveBeenCalledWith(props.min);
    });
});
