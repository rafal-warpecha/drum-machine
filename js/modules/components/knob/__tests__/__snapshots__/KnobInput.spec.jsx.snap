// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KnobInput component should match snapshot 1`] = `
<input
  className="input"
  max={100}
  min={0}
  onBlur={[Function]}
  onChange={[Function]}
  type="number"
  value={0}
/>
`;
